tchannel/tornado/connection.py

```diff
@@ -452,12 +466,26 @@ def serve(self, handler):
         """
         assert handler, "handler is required"
         assert self._loop_running, "Finish the handshake first"
-
         while not self.closed:
             message = yield self.await()
-
             try:
-                handler(message, self)
+                if self.draining:
```

Comment:
Look at how many levels of indentation and branching there are! Something is wrong.

When you're shutting down why not just remove all of your handlers and replace your fallback handler with one that returns "declined"?